### ISSUE TITLE
Memory leak in join node

### ISSUE THREAD

**Issue by natcl** [node-red/node-red#4080]
### Current Behavior

When using the `split` and `join` nodes to split an `Array` that then fetches an `Array` and gets fed back to a `join` node, the join node will accumulate messages in the `inflight` object and won't clear those messages. This causes a memory leak where memory will grow over time and the objects will never be cleared.

<img width="1317" alt="Capture d’écran, le 2023-03-01 à 10 31 40" src="https://user-images.githubusercontent.com/77556/222186528-7477751d-9d48-4568-979e-ada769d74ca6.png">

### Expected Behavior

We expect the join node to clear the inflight object.

### Steps To Reproduce

See example flow.
To aid debugging I added the following line before [this](https://github.com/node-red/node-red/blob/604487143886cd5a5150f269e5d244934df29fcc/packages/node_modules/%40node-red/nodes/core/sequence/17-split.js#L576) line in 17-split.js : 
```
node.warn(Object.keys(inflight).length)
```

### Example flow

```
[{"id":"2dfa31275516be36","type":"comment","z":"f592b8e583339c68","name":"Each time we click the inject, it seems fine but the inflight object in the join node is not cleared and accumulates indefinetly","info":"","x":460,"y":100,"wires":[]},{"id":"2c9831e66b101a52","type":"comment","z":"f592b8e583339c68","name":"To test this I added node.warn(Object.keys(inflight).length) after line 573 of the 17-split.js file","info":"","x":360,"y":140,"wires":[]},{"id":"2239d4bbf0c0e0e7","type":"group","z":"f592b8e583339c68","name":"We create 2 arrays","style":{"label":true},"nodes":["2ee2ffe044536db6","1261dcade049d5eb"],"x":34,"y":299,"w":392,"h":82},{"id":"2ee2ffe044536db6","type":"template","z":"f592b8e583339c68","g":"2239d4bbf0c0e0e7","name":"animals","field":"animals","fieldType":"flow","format":"handlebars","syntax":"mustache","template":"cats:\n  - name: steve\n    age: 33\n  - name: stevette\n    age: 32\ndogs:\n  - name: boumbo\n    age: 33\n  - name: bamba\n    age: 32","output":"yaml","x":340,"y":340,"wires":[[]]},{"id":"1261dcade049d5eb","type":"inject","z":"f592b8e583339c68","g":"2239d4bbf0c0e0e7","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":true,"onceDelay":0.1,"topic":"","payload":"","payloadType":"date","x":150,"y":340,"wires":[["2ee2ffe044536db6"]]},{"id":"e7908d8fae1bb759","type":"group","z":"f592b8e583339c68","name":"","style":{"label":true},"nodes":["b91186b9d6c5d3e0","00e02e1d2d788223","e9fed81e08f5880a","8951a9e4b30cf306","100462d6532aa4c0"],"x":34,"y":179,"w":1142,"h":82},{"id":"b91186b9d6c5d3e0","type":"inject","z":"f592b8e583339c68","g":"e7908d8fae1bb759","name":"","props":[{"p":"payload"},{"p":"topic","vt":"str"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","payload":"[\"cats\", \"dogs\"]","payloadType":"json","x":160,"y":220,"wires":[["00e02e1d2d788223"]]},{"id":"00e02e1d2d788223","type":"split","z":"f592b8e583339c68","g":"e7908d8fae1bb759","name":"","splt":"\\n","spltType":"str","arraySplt":1,"arraySpltType":"len","stream":false,"addname":"","x":360,"y":220,"wires":[["100462d6532aa4c0"]]},{"id":"e9fed81e08f5880a","type":"join","z":"f592b8e583339c68","g":"e7908d8fae1bb759","name":"","mode":"auto","build":"object","property":"payload","propertyType":"msg","key":"topic","joiner":"\\n","joinerType":"str","accumulate":true,"timeout":"","count":"","reduceRight":false,"reduceExp":"","reduceInit":"","reduceInitType":"","reduceFixup":"","x":900,"y":220,"wires":[["8951a9e4b30cf306"]]},{"id":"8951a9e4b30cf306","type":"debug","z":"f592b8e583339c68","g":"e7908d8fae1bb759","name":"debug 43","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","statusVal":"","statusType":"auto","x":1070,"y":220,"wires":[]},{"id":"100462d6532aa4c0","type":"change","z":"f592b8e583339c68","g":"e7908d8fae1bb759","name":"get correct array according to payload","rules":[{"t":"set","p":"payload","pt":"msg","to":"animals[msg.payload]","tot":"flow"}],"action":"","property":"","from":"","to":"","reg":false,"x":640,"y":220,"wires":[["e9fed81e08f5880a"]]}]
```

### Environment

- Node-RED version: 3.0.2
- Node.js version: 18
- npm version: 
- Platform/OS: MacOS, Alpine, Ubuntu
- Browser:

**Comment by dceejay** [comment]
In theory the inflight should be cleaned up if in array mode and the array is complete. This is so that ( in theory) multiple partial arrays could be being filled up in parallel, each only being sent when complete. I guess this isn't happening correctly. 
( away this week so can't dig in )

**Comment by knolleary** [comment]
The logic to clear out the inflight messages only happens if the `accumulate` option is not set. Whilst that is true for manual mode, in automatic mode that should happen automatically.

The workaround is:

1. edit the join node
2. set to 'manual' mode
3. untick the 'And Every Subsequent Message' option (this will set `node.accumulate` to false)
4. set back to 'automatic' mode
5. save changes and deploy

**Comment by natcl** [comment]
Thanks for the workaround ! I was sending `msg.reset` after my join but the workaround is cleaner.
Is that still something that needs fixing in the join node ?
I feel that it's quite easy for someone to get caught by this as the split/join combo is often used in automatic mode.

**Comment by knolleary** [comment]
Fixed via #4083 in 3.1

**Comment by natcl** [comment]
Oh that answers my question then :)
Thanks for the fix !
